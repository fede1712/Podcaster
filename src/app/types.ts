export interface Podcast {
  "im:name": { label: string };
  "im:image": image[];
  summary: { label: string };
  "im:price": { label: string; attributes: { amount: string; currency: string } };
  "im:contentType": { attributes: { term: string; label: string } };
  rights: { label: string };
  title: { label: string };
  link: { attributes: { rel: string; type: string; href: string } };
  id: { label: string; attributes: { "im:id": string } };
  "im:artist": { label: string; attributes: { href: string } };
  category: { attributes: { "im:id": string; term: string; scheme: string; label: string } };
  "im:releaseDate": { label: string; attributes: { label: string } };
}

export type image = { label: string; attributes: { height: string } };

export interface Episode {
  country: string;
  artworkUrl160: string;
  episodeFileExtension: string;
  episodeContentType: string;
  artworkUrl600: string;
  collectionViewUrl: string;
  trackTimeMillis: number;
  shortDescription: string;
  closedCaptioning: string;
  collectionId: number;
  collectionName: string;
  genres: genres[];
  episodeGuid: string;
  description: string;
  releaseDate: string;
  previewUrl: string;
  trackId: number;
  trackName: string;
  feedUrl: string;
  artistIds: [];
  episodeUrl: string;
  artworkUrl60: string;
  contentAdvisoryRating: string;
  trackViewUrl: string;
  kind: string;
  wrapperType: string;
  genreIds: string[];
  primaryGenreName: string;
  currency: string;
  trackCount: number;
  trackExplicitness: string;
  artistName: string;
  collectionCensoredName: string;
  trackCensoredName: string;
  artworkUrl30: string;
  artworkUrl100: string;
  collectionPrice: number;
  trackPrice: number;
  collectionHdPrice: number;
  collectionExplicitness: string;
}

export type genres = { name: string; id: string };
