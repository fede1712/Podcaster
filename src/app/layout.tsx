import type { Metadata } from "next";
import { Inter } from "next/font/google";
import "./globals.css";
import Link from "next/link";
import { Header } from "./ui/header/header";

export const metadata: Metadata = {
  title: "Podcaster",
  description: "Itunes top podcasts",
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body className="m-8">
        <Header />
        {children}
      </body>
    </html>
  );
}
