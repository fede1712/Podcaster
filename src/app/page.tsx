import { getPodcasts } from "./lib/data";
import { PodcastCard } from "./ui/podcastCard/podcastCard";

export default async function Home() {
  const data = await getPodcasts();
  console.log(data.feed.entry[1]);

  return (
    <div className="grid grid-cols-4 gap-14 m-14">
      {data.feed.entry.map((podcast: any) => {
        return <PodcastCard author={podcast["im:artist"].label} title={podcast["im:name"].label} src={podcast["im:image"][2].label} />;
      })}
    </div>
  );
}
