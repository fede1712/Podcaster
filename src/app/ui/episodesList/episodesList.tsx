import { Episode } from "@/app/types";
import Link from "next/link";

export const EpisodesList = ({ data, podcastId }: { data: { results: Episode[] }; podcastId: string }) => {
  return (
    <table className="table-auto border shadow-md mt-10 p-10 w-full" data-testid="episode-list">
      <thead>
        <tr className="grid grid-cols-10 justify-items-stretch w-full p-4 border-b-2">
          <th className="col-span-6 justify-self-start">Title</th>
          <th className="col-span-2 justify-self-end">Date</th>
          <th className="col-span-2 justify-self-end">Duration</th>
        </tr>
      </thead>
      <tbody className="w-full m-10">
        {
          (data.results.shift(),
          data.results.map((episode: Episode, idx: number) => {
            const date = new Date(episode.releaseDate);
            const duration = new Date(episode.trackTimeMillis);
            const episodeId = episode.trackId;
            return (
              <tr key={idx} className="grid grid-cols-10 justify-items-stretch w-full p-2  border-b-2">
                <td className="text-cyan-600 col-span-6 justify-self-start">
                  <Link href={`/podcast/${podcastId}/episode/${episodeId}`}>{episode.trackName}</Link>
                </td>
                <td className="col-span-2 justify-self-end">{`${date.getDate()} / ${date.getMonth() + 1} / ${date.getFullYear()}`}</td>
                <td className="col-span-2 justify-self-end">{`${duration.getMinutes()}:${duration.getSeconds()}`}</td>
              </tr>
            );
          }))
        }
      </tbody>
    </table>
  );
};
