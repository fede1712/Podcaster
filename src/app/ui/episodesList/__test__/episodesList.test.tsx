import React from "react";
import { render, screen } from "@testing-library/react";
import { Episode } from "@/app/types";
import { EpisodesList } from "../episodesList";

test("Renders episodes list correctly", () => {
  const episodeList: Episode[] = [
    {
      trackName: "Test Episode 1",
      description: "This is a test episode description. 1",
      episodeUrl: "http://example.com/test-episode.mp3-1",
      artworkUrl600: "http://example.com/artwork.jpg-1",
      trackTimeMillis: 123456,
      artistName: "Test artist name 1",
      collectionName: "Test collection name 1",
      releaseDate: "Test realease date 1",
      trackId: 1,
    },
    {
      trackName: "Test Episode 2",
      description: "This is a test episode description. 2",
      episodeUrl: "http://example.com/test-episode.mp3-2",
      artworkUrl600: "http://example.com/artwork.jpg-2",
      trackTimeMillis: 123456,
      artistName: "Test artist name 2",
      collectionName: "Test collection name 2",
      releaseDate: "Test realease date 2",
      trackId: 2,
    },
  ];

  render(<EpisodesList data={{ results: episodeList }} podcastId="123456" />);
  const tableTitleElement = screen.queryByText("Title");
  expect(tableTitleElement).not.toBeNull();

  const tableDateElement = screen.queryByText("Date");
  expect(tableDateElement).not.toBeNull();

  const tableDurationElement = screen.queryByText("Duration");
  expect(tableDurationElement).not.toBeNull();

  const { asFragment } = render(<EpisodesList data={{ results: episodeList }} podcastId="123456" />);

  expect(asFragment()).toMatchSnapshot();
});
