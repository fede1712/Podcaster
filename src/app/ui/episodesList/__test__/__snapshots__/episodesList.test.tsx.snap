// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders episodes list correctly 1`] = `
<DocumentFragment>
  <table
    class="table-auto border shadow-md mt-10 p-10 w-full"
  >
    <thead>
      <tr
        class="grid grid-cols-10 justify-items-stretch w-full p-4 border-b-2"
      >
        <th
          class="col-span-6 justify-self-start"
        >
          Title
        </th>
        <th
          class="col-span-2 justify-self-end"
        >
          Date
        </th>
        <th
          class="col-span-2 justify-self-end"
        >
          Duration
        </th>
      </tr>
    </thead>
    <tbody
      class="w-full m-10"
    />
  </table>
</DocumentFragment>
`;
